### PR TITLE
Implement support for element navigation

### PR DESCRIPTION
**Pull request by lforst**
Part-of eclipse-glsp/glsp#204

### CODE REVIEW

File: vscode-integration/src/action/operation.ts [hunk 1]
```diff
@@ -46,3 +46,13 @@ export class LayoutOperation implements Operation {
             && 'elementIds' in action;
     }
 }
+
+export class NavigateAction implements Operation {
+    static readonly KIND = 'navigate';
+    readonly kind = NavigateAction.KIND;
+    constructor(readonly targetTypeId: string) { }
```
Comment (tortmayr) [inline vscode-integration/src/action/operation.ts:53]: This definition is missing the `args` property (See also https://github.com/eclipse-glsp/glsp-client/blob/4e7da984e79b162cf4c5ccc3cedf82ced047ddb9/packages/client/src/features/navigation/navigation-action-handler.ts#L53)

File: vscode-integration/src/glsp-commands.ts [hunk 2]
```diff
@@ -32,6 +32,9 @@ export namespace GLSPCommand {
     export const CENTER = 'center';
     export const DELETE = 'delete';
     export const LAYOUT = 'layout';
+    export const NEXT_NODE = 'nextNode';
```
Comment (tortmayr) [inline vscode-integration/src/glsp-commands.ts:35]: The navigation commands are diagram-specific and should be defined in the workflow-example package.

File: vscode-integration/src/glsp-diagram-editor-provider.ts [hunk 3]
```diff
@@ -59,6 +59,24 @@ export class GlspDiagramEditorProvider implements vscode.CustomEditorProvider<Gl
             command: GLSPCommand.LAYOUT,
             ...options
         });
+
+        GLSPCommand.registerActionCommand({
```
Comment (tortmayr) [inline vscode-integration/src/glsp-diagram-editor-provider.ts:63]: The registration of the diagram-specific navigation commands should be done in `WorkflowGlspDiagramEditorContext`.
Maybe we should move the `registerCommands` method from the `EditorProvider` to the `GlspDiagramEditorContext`.

**Review by tortmayr — CHANGES_REQUESTED** [review]
Thanks for providing this draft.
For the show-documentation navigation case you need to define a `NavigateToExternalTargetActionHandler` (see also https://github.com/eclipse-glsp/glsp/issues/204#issuecomment-822375947)

I had a quick look at you code and left some pointers/suggestions as inline comments

File: example/workflow/extension/package.json [hunk 4]
```diff
@@ -44,6 +63,24 @@
         "title": "Layout diagram",
         "category": "Workflow Diagram",
         "enablement": "glsp-workflow-diagram-focused"
+      },
+      {
+        "command": "workflow.glsp.diagram.nextNode",
```
[the remark this answers is not in the document]
Comment (lforst) [reply example/workflow/extension/package.json]: I will put this off for now as I believe that it requires digging deeper into diagram state context for the `when` clause.

File: vscode-integration/src/action/operation.ts (hunk 1 above)
Comment (lforst) [reply vscode-integration/src/action/operation.ts]: Hi, I added the args property.

File: vscode-integration/src/glsp-commands.ts (hunk 2 above)
Comment (lforst) [reply vscode-integration/src/glsp-commands.ts]: I did my best trying to extract the registration of commands and their actions into user space. Feel free to feedback whether this is the direction we want to move in.

Comment (tortmayr) [reply vscode-integration/src/glsp-commands.ts]: > I am a bit stuck regarding the external navigation. For some reason the `NavigateToExternalTargetAction` never gets handed through to a place where we can interact with the VSCode api.
> 
> The only way I found to intercept the Action is to bind an action handler in the `GLSPStarter` via the `configureActionHandler` method (See [dacd7af](https://github.com/eclipse-glsp/glsp-vscode-integration/commit/dacd7afed4257911ad6c68c476ecfa2c4387e8b6)). The problem now I believe is, that we are in the webview space and can't access the vscode api from inside. When I try to import `vscode` in the action handler I get a transpilation error that the module can't be found (even though I installed). I believe I am missing something obvious here...
> I would be very greatful if you had any ideas on how to solve this.

You are right, this looks like a major limitation in our action handling flow. There is currently no possibility to handle a local webview action (i.e. a action that was dispatched from within glsp/sprotty and is not delegated to the glsp server) in the vscode extension. However, the vsode extension is the only place where you have access to the vscode API. As you already noticed you cannot use the vscode API inside the webview, even if you import it. I have created https://github.com/eclipse-glsp/glsp/issues/234 to track this and I'm already working on a fix.

Comment (lforst) [reply vscode-integration/src/glsp-commands.ts]: Thanks for looking into this! I'm glad my attempts to get this working were not completely in vain.

Comment (tortmayr) [reply vscode-integration/src/glsp-commands.ts]: I have open a [PR](https://github.com/eclipse-glsp/glsp-vscode-integration/pull/6) to fix this #6. I tested this with a mock `ExtensionActionHandler` for the `NavigateToExternalTargetAction` and everything seems to work as expected now.

File: example/workflow/workspace/example1.wf [hunk 5]
```diff
@@ -21,11 +21,11 @@
                   "type": "label:icon",
                   "position": {
                     "x": 7.65625,
-                    "y": 5.0
+                    "y": 4.5
                   },
```
Comment (tortmayr) [inline example/workflow/workspace/example1.wf:25]: Looks to me like these changes were committed on accident?

**Review by tortmayr — COMMENTED** [review]
Code looks good to me and the `Show Documentation` use case works like a charm for me. Unfortunately I have troubles with the `go to` node use case. This only works the first time a execute a one of the goto commands. Subsequent command executions will no longer work. If I close and reopen the diagram I can successfully execute a go-to command one time again.

**Comment by lforst** [comment]
> Code looks good to me and the `Show Documentation` use case works like a charm for me. Unfortunately I have troubles with the `go to` node use case. This only works the first time a execute a one of the goto commands. Subsequent command executions will no longer work. If I close and reopen the diagram I can successfully execute a go-to command one time again.

Thanks for taking a look at this! :)

The behaviour you're describing is very interesting. Sadly I can't reproduce this error. What I've found is that you can't call any navigation commands on any type of nodes besides `Automated Tasks` and `Manual Tasks`. Might this be the reason the commands did not work for you?

**Comment by lforst** [comment]
Is there an update on the `Show Documentation` command that didn't work for you? @tortmayr

Other than that is there a reason that prevents this PR from being merged? I'm ready to implement suggestions of any kind.

**Comment by tortmayr** [comment]
I cannot reproduce the behavior anymore so from my side there is nothing that prevents merging this. However, there is currently a merge conflict. Could you please resolve the conflict in  `packages/vscode-integration/src/action/index.ts`

**Comment by lforst** [comment]
I resolved the merge conflict and everything should be good to go. Thanks again for taking the time to take a look at my changes!

**Review by tortmayr — APPROVED** [review]
:+1:

**Pull request merged**